### ISSUE TITLE
Question: Optimization techniques used by Pyston

### ISSUE THREAD

**Issue by ssili** [pyston/pyston_v2#238]
I compared CPython3.8.12 with Pystonv2.2 on some benchmarks and found that Pyston was faster than CPython. The reason I use Pystonv2.2 instead of the latest version is that I tried it out as soon as it was released. 

I wondered why Pyston is so fast, so I checked the source code of Pystonv2.2. I summarize some reasons based on [comments](https://github.com/pyston/pyston/blob/43fd0d3dcdf3999e8c53fb51e5cce77424d5bd36/Python/aot_ceval_jit.c#L1) in `aot_ceval_jit.c` file and [wiki](https://github.com/pyston/pyston/wiki/Our-techniques). The main reasons I summarize are as follows, but I don't know if there is any error.

- JIT: JIT the interpreter loop into machine code (mentioned in wiki), which removes the dispatching overhead of the interpreter.
- Aggressive attribute caching (mentioned in [README.md](https://github.com/pyston/pyston#readme)).
- `Deferred value stack`: reduce unnecessary reference count, reduce overhead related to the manipulation of Python value stack, remove many memory operations.
- Specialization: generate some specialized function. For example, some functions end with `Profile`, determine types of the operands, and call specific functions (as [commented](https://github.com/pyston/pyston/blob/43fd0d3dcdf3999e8c53fb51e5cce77424d5bd36/Python/aot_ceval_jit.c#L8) in `aot_ceval_jit.c`).
- Some small optimizations throughout the CPython codebase controlled by macro `PYSTON_SPEEDUPS`.

Is there any other major technique used by Pyston to improve the performance? 

I decide to investigate how much these optimization techniques contribute to the performance of Pyston by comparing time and memory consumption, and some metrics provided by perf, such as the amount of load/store events, context switch, branches, etc. So, I want to disable some optimization techniques used in Pyston by modifing the source code of Pystonv2.2. Is there any better way to perform this analysis?

I want to modify the source code of **Pystonv2.2** to disable some optimization techniques in the following way, but I don't know **whether it is correct** or there is any better way.

- disable JIT: set the macro `ENABLE_AOT` in [`pylifcecycle.c`](https://github.com/pyston/pyston/blob/43fd0d3dcdf3999e8c53fb51e5cce77424d5bd36/Python/pylifecycle.c#L605) file. 
- disable caching mechanism: modify the `emit_inline_cache` function to always return `1` and set the `opcache_available` argument  in `get_aot_func_addr` function to be 0. When Pyston run in interpreted execution mode, set related macro, such as `USE_LOAD_METHOD_CACHE`.
- disable optimization provided by `deferred value stack`: set related macro, such as `ENABLE_DEFERRED_LOAD_FAST` or set the capacity of deferred value stack to be 0.
- disable specialized function: set `jit_use_aot` to be 0.

**Comment by undingen** [comment]
Thanks for the detailed interest in Pyston, do you plan to release the results of your investigations? I would also be very interested and what you find out. @kmod did a similar but less detailed investigation recently so he may have more info. 

I think everything your wrote is correct but I would suggest using the environment variables to disable different stuff (but be careful pyperf per default filters the env vars). Because running `make opt` for every change will take a long time.
- to disable the JIT: `JIT_MIN_RUNS=9999999999999` (will only JIT after that many function calls / loop iterations - counter is a 64bit long)
- disable the inline caches for `LOAD_ATTR`, `LOAD_METHOD` etc,: `JIT_USE_ICS=0`
- disable calling into the type specialized paths for bytecodes: `JIT_USE_AOT=0`
- to disable deferred value stack you will need to add your own environment variable or use the the C macros like you mentioned.

Also I guess some of the optimization depend on others which makes it hard to measure things. E.g. if you disable the "deferred value stack" it may also disable some inline caches etc..

**Comment by kmod** [comment]
I agree with Marius, this sounds really interesting! As he mentioned a number of our optimizations are interrelated so sometimes the results are hard to interpret, and on a practical note we don't test the reduced-functionality versions of Pyston so sometimes there are bugs / things not fully disabled by the environment variables. But anyway I'm very interested to see what you find!
